sycl/doc/extensions/proposed/sycl_ext_intel_virtual_functions.asciidoc

```diff
@@ -0,0 +1,579 @@
+  // Exception is expected to be thrown, because target device doesn't support
+  // fp64 aspect and it is required by 'Foo::foo' which is included into 'set_A'
+});
+
+q.single_task(syclext::properties{syclext::calls_indirectly<set_B>}, [=] {
+  // No exceptions are expected, because 'set_B' doesn't bring any requirements
+  // for optional kernel features.
+});
+----
+
+An implementation may not raise a compile time diagnostic or a run time
+exception merely due to speculative compilation of a virtual member function for
+a device when the application does not specify a use of virtual member functions
+through the correponding properties.
+
+[source,dpcpp]
+----
+using syclext = sycl::ext::oneapi::experimental;
+
+struct Foo {
+virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY() void foo() {
+  double d = 3.14;
+}
+
+virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY() void bar() {}
+};
+
+sycl::queue q(/* device selector choosing a device *without* fp64 support */);
+assert(!q.get_device().has(sycl::aspect::fp64));
+
+auto *Storage = sycl::malloc_device<Foo>(1, q);
+
+q.single_task([=] {
+  // The kernel is not submitted with 'calls_indirectly' property and therefore
+  // it is not considered to be using any of virtual member functions of 'Foo'.
+  // This means that the object of 'Foo' can be successfully created by this
+  // kernel, regardless of whether a target device supports 'fp64' aspect which
+  // is used by 'Foo::foo'.
+  // No exceptions are expected to be thrown.
+  new (Storage) Foo;
+});
+----
+
+==== Interaction with `reqd_sub_group_size` attribute
+
+The `reqd_sub_group_size` attribute is a bit of a special case comparing to
+other optional kernel features, because it requires to compile a kernel in a
+certain way, which may require special handling for all functions which are
+called from it.
+
+When the same helper function is called from two or more kernels
+with different `reqd_sub_group_size` attribute, it may be required for the
+implementation to duplicate that helper function to create different versions
+of it tailored to different sub-group sizes. It can be done in a straightforward
+manner when operating on a static call graph.
+
+Virtual member functions are essentially called indirectly and pointers to them
+are initialized just once when an object of a polymorhpic class is being
+created. Therefore, to support calling such virtual member function from two or
+more kernels with different `reqd_sub_group_size`, each kernel may need to
+receive a different pointer to a different version of a virtual member function.
+
+To avoid possibly posing such multi-versioning requirements on implementations,
+virtual member functions can only be called from kernels with _primary_
+sub-group-size as defined by
+link:../proposed/sycl_ext_oneapi_named_sub_group_sizes.asciidoc[
+sycl_ext_oneapi_named_sub_group_sizes] extension, or otherwise behavior is
+undefined.
```

Comment:
Ok, my concern still stands: because we don't have a way to set reqd-sub-group size for a non-kernel function at SPIR-V level, we can't really annotate them at SYCL level. This is more of an implementation issue than a language issue, but I've recorded it in 95fd59fa592e9c66a7c3352d9771252130f68d04 to improve the situation later.